### ISSUE TITLE
Revisit custom editor input factory

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#117208]
While working on https://github.com/microsoft/vscode/pull/116856 I spent some though on the custom editor input factory. The only usecase today seems to be for restoring backups (via [`BackupRestorer`](https://github.com/microsoft/vscode/blob/a830cb5f74e5a487dd8a764c84e1fa38f3e763dd/src/vs/workbench/contrib/backup/common/backupRestorer.ts#L22-L22)). 

I would think that a better place for this would be `IEditorService#createEditorInput` and the backup restorer should just use that. If we need to make this method async for a few usecases I think that is OK, as long as the signature could stay sync for most typical uses.

**Comment by bpasero** [comment]
//cc @jrieken

**Comment by jrieken** [comment]
related to https://github.com/microsoft/vscode/issues/117899#issuecomment-788056636

**Comment by jrieken** [comment]
and related to https://github.com/microsoft/vscode/issues/117873

**Comment by jrieken** [comment]
Yes, I have to say that I can't wrap my head around this. I don't understand why notebooks must register a `IEditorInputFactory` _and_ `ICustomEditorInputFactory` and I am failing to fix #117899 because not of this is intuitive to me. 

Isn't the idea of an editor input to abstract away editors? And isn't an input factory a way to revive a serialised input into an alive object? That sounds a like pretty good abstraction and something must be very wrong with it if we now require the same abstraction again, just for custom editors.

**Comment by jrieken** [comment]
Marking all of this as important because this is crucial to the notebooks effort and playing the catch-up game of copy-pasting source code isn't sustainable.

**Comment by lramos15** [comment]
@bpasero We should probably set up another meeting to discuss this as I don't think there's a good solution here. After discussion we weren't able to merge these ideas.

**Comment by bpasero** [comment]
Yeah I think this deserves some more discussion with custom editor + notebook people why the `viewType` is needed so early on for the editor input to work. 

To follow up on @jrieken point:

> Isn't the idea of an editor input to abstract away editors? And isn't an input factory a way to revive a serialised input into an alive object? That sounds a like pretty good abstraction and something must be very wrong with it if we now require the same abstraction again, just for custom editors.

Yes exactly, an editor input should be sufficient to open an editor and it shouldn't require an `async` lookup into backups. The flow should be:
* editor input gets created (either from a backup that was found or because it was opened previously)
* editor opens (text, notebook or custom)
* in the `resolve` method you can do the actual lookup of what view type is needed

One thing that is special here is that the need for the custom editor input factory only applies to one case:
* you open a folder and make editors dirty (text, custom, notebooks)
* you close that window and exit without saving
* from the command line you type `code <folder> <file>`
* we only open the `file` and do not restore the other editors
* but we still of course want to restore all editors that are dirty so we need this complicated flow of knowing how to actually create the inputs

@lramos15 maybe we can actually solve this differently in the end: if an editor ends up as backup it must be opened and as such we must have it as UI state somewhere. Instead of doing this complicated dance with the custom editor input factory, we can try to get the editor from that UI state instead. We can also think about simply restoring all previously opened editors as soon as we detect any dirty editor. In any case we can probably get rid of the custom editor input factory entirely 👍

**Comment by lramos15** [comment]
@jrieken I'm removing the important label as I think we have at least blocked the catchup game with the new `editorOverrideService`.

@bpasero Do you still want to try to tackle this in April? I'm not sure how important it is. It doesn't really improve the usability of the editorOverrideService from my understanding as we never operate on the backup case.

**Comment by bpasero** [comment]
Will update here once I looked into how to restore backups with the new `typeId` support.
